File: Array/200.Number_of_Islands/Number_of_Islands.cpp
class Solution {
public:
    void DFS(vector<vector<char>>& grid, int i, int j){
        grid[i][j] = '$';
        
        
        if(i > 0 && grid[i - 1][j] == '1') DFS(grid, i - 1, j);

        if(i < grid.size() - 1 && grid[i + 1][j] == '1') DFS(grid, i + 1, j);

        if(j > 0 && grid[i][j - 1] == '1') DFS(grid, i, j - 1);

        if(j < grid[i].size() - 1 && grid[i][j + 1] == '1') DFS(grid, i, j + 1);

    }
    
    
    
    
    int numIslands(vector<vector<char>>& grid) {
        
        int ans = 0;
        
        for(int x = 0; x < grid.size(); x++){
            
            for(int y = 0; y < grid[x].size(); y++){
                
                if(grid[x][y] == '1'){
                    ans++;
                    DFS(grid, x, y);
                }
            
            }
        }
        return ans;
    }
};

File: Array/200.Number_of_Islands/Number_of_Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ans = 0
        
        
        def dfs(i, j):
            grid[i][j] = "$" 
    
            if i > 0 and grid[i - 1][j] == "1": dfs(i - 1, j)
            if i < len(grid) - 1  and grid[i + 1][j] == "1":  dfs(i + 1, j)
            if j > 0  and grid[i][j - 1] == "1": dfs(i, j - 1)
            if j < len(grid[i]) - 1 and grid[i][j + 1] == "1": dfs(i, j + 1)
        
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    ans += 1
                    dfs(i, j)
                    
        
        return ans